### DISCUSSION TITLE
Ctrlr Lua Version (confirming 'built-in' & whether Ctrlr survives attempts to use newer version)

### DISCUSSION THREAD

**Discussion by unityconstruct** [RomanKubiak/ctrlr#599] [Q&A] [answered]
Hey CTRLR Fam,

Curious if
1. the CTRLR Lua IDE environment is 'loosely bound' such that a newer LUA version can be used
2. if a newer version is possible, assuming a rebuilt is required
3. and that cpp code needs to be updated to include it ( ie: not controller by config files that `make` might auto-tasticly handle.). Given CTRLR's portability - near certain it needs headers updated, libs pulled in, recompile.

**Really just checking pulse on if its more trouble than its worth.. so:**

- Poked around the codebase and found some includes.. My background is in java/C#(plus a stupid number of others of varying profeciency)  & they are quite the departure from their cpp progenitor's ecosystem. 
- lua.h has a `LUA_VERSION` so just confirming that one can safely assume that CTRLR natively uses this library for compilation ( thus IDE compiling
  - if I CAN'T pull in a newer version _(5.4.6 is latest it seems)_, then I reason its best to jump into *this* version's reference & train up
  - if I CAN pull in a newer version, wasn't sure on my cursory look if its as simple as updating the specs here.
  - If its convoluted & documented, I'd like to struggle through that before wasting folks' time here on the minutia.

```output
ctrlr/Source/Misc/lua/include
** $Id: lauxlib.h,v 1.88.1.1 2007/12/27 13:02:25 roberto Exp $

/lua.h
** $Id: lua.h,v 1.218.1.5 2008/08/06 13:30:12 roberto Exp $
#define LUA_VERSION	"Lua 5.1"
#define LUA_RELEASE	"Lua 5.1.4"

** $Id: luaconf.h,v 1.82.1.7 2008/02/11 16:25:08 roberto Exp $
** $Id: lualib.h,v 1.36.1.1 2007/12/27 13:02:25 roberto Exp $
```

Thanks in advance for any responses...
AND... **hats off to RK** for producing this amazing tool & **equal accolades to the chaps here keeping it all on life support**. This is one of those projects I feel strongly about & hope I can get around to learning CPP/LUA/JUIC enough to make updates  _(if only for additional panels)_ before the SMEs are lost to time. Codebase looks succinct, but it might deceptive due to gorrilion-line-classes _(many of which I am credited with also)_. CTRLR reminds me of FruityLoop v 0.93... gasping & nearly dead for years, then some critical mass hits & its ubiquitous with music production and is one of the most sophisticated UIs I've dealt with.

Cheers!
UC

**Comment by unityconstruct** [comment]
# P.S.: **VS: build-ctrlr-on-windows-10-manual**

## Might as well & cross-pollinate my topic as it might address my noobQ(s): 
- _**Note:** While I have built CTRLR on Linux already last year & don't think I bothered with involving VS at the time._
- #592:
  - Was parsing board & saw reference to a `VS build doc` ~2 weeks ago
- #99:
  - Found another post with the link @dobo365 mentioned, but looks like it might be behind the elder-forum `file-access-perms` 
  - not sure if there's generally accessible URI for this guy
- Archive.org: Can't fathom there would be, but figured I ask:
  - For esoteric stuff like this I tend to push things to archive.org for posterity - and on a bad day, I can still get to it if I fail to organize before I lose it. I know there is some level of IP moving across the boards, so wanted to ensure I don't violate any protocol with the habit. 
  - Case in point: witnessed the `deep-sixing` of some fabulous panel work & I'd imagine this entails obscene levels of LUA/implementation examples.


#592 
[dobo365](https://github.com/dobo365)
[2 weeks ago](https://github.com/RomanKubiak/ctrlr/discussions/592#discussioncomment-8099737)
Author
Thx a lot @bijlevel ! Your **Word document describing how to build Ctrlr** and the good advices from Damien helped me to understand how to build. I have been looking a lot at the code and start to know what is where (I said "start").
Unfortunately, for the initial changes I wanted to do I faced this CtrlrIDs.xml issue...

FYI, I just bought a Mac M1 second hand so will need your M1 build and would also like how to build Ctrlr on Mac (will also take that with Damien)

<hr/>

## POST WITH ATTACHMENT

- https://ctrlr.org/forums/topic/build-ctrlr-on-windows-10-manual/
- October 11, 2020 at 10:42 am[#120246](https://ctrlr.org/forums/topic/build-ctrlr-on-windows-10-manual/#post-120246)
<br/>
"""
<br/>
[Tedjuh](https://ctrlr.org/forums/users/tedjuh/)
Participant
Topics: 9
Replies: 97
Total: 106
★★
As promised a manual (attachment) to get Ctrlr from the Ctrlr repository and instructions on how to compile/ build it successfully. It might not be complete. I might have forgotten one or two steps but it should give you a headstart.

Questions or additions? No PM’s or e-mails, please. Ask or put them here. We might all learn from it.

In need of VS 2010?. PM me and I’ll see how the Compile-Gods react to your humble phrased question.

And no, I’m not planning on doing it for Mac-OS nor Linux.

Attachments:
You must be [logged in](https://ctrlr.org/wp-login.php?redirect_to=https%3A%2F%2Fctrlr.org%2Fforums%2Ftopic%2Fbuild-ctrlr-on-windows-10-manual%2F) to view attached files.
"""

_Thanks in advance for any resposes & sorry for comment length!_

**Comment by Ted-juh** [comment]
I wrote that manual using VS2010 because it would be less hassle to build Ctrlr 5.3.201 with the tools that were probably closest to the one that Roman might have used. I remember I tried to compile Ctrlr 5.3.201 with a newer version of VS but ran into a couple of errors that I couldn't resolve probably due to inexperience. 

Bylevel's manual is for Ctrlr 5.6 if I remember correctly which uses Juce 6.08. An excellent manual indeed.

About the Lua subject. It's not an easy task to upgrade the Lua version just because the boost library it depends on to do the Luabind hasn't been maintained for years, the same goes for Luabind. So, if you want to update the Lua version, you might as well look at Sol2 (v3) or Luajit to do the binding to C++ without the need for Boost.

The project has a stubborn (sorry guys :)) club of members. Synth lovers who like their machines to be usable on today's computers. Stubborn in a way that we want to revive what was once almost dead. Is it dead? No. Le Roi est mort, vive le Roi.

**Reply by unityconstruct** [reply]
**Hopefully, this injects a spirit of positivity**
**FruityLoops 1.5.23 in 1998**
- This app was novel, but nearly dead on arrival with max 4 wavs to loop & in 'The Spirit of RubberDuck H20'
- Using it now its hard to fathom 1. that it sucked that bad, 2. crashed that much for so little, 3. having survived the extinction-level event it revolutionized music production for the common man & inspired a ton of shops to compete with it.
- Conceptually `CTRLR` is on 'that level', but with the possibilities provided by open source(note linux in the face of snearing AV peeps once)  & the fact it extends the shelf life & usefulness of tons of antique gear constantly getting eyed for the trash heap means something.
- After procuring the X-Station panel.. AND it being VST.. I was immediately sold.. I'm over here eyeballing my E-MU collection of SysExMania with lust.. knowing I'm going to have to date the CTRLR for a year or two before it gives me the keys to that room :P
- ok enough nostalgia/spamming the boards.. thx again!

![image](https://github.com/RomanKubiak/ctrlr/assets/40268430/8ec31ad1-625e-4ea8-9077-3ee3a1d6b5a4)

**Comment by unityconstruct** [comment]
- haha... Sweet.. EXACTLY the info I was needing & expecting to hear...
- I just hopped on the LUA dotorg & had that feeling if I dive off into 5.4.whateves... I'd have the opportunity to do it all over again for 5.1.whatevs
- 

**Observations:**
__"The project has a stubborn (sorry guys :)) club of members. "__
THIS BE LIKE: The project has an ~~stubborn (sorry guys :))~~ **unpaid** club of members
- the last guy that touches codebase is the first to require witness protection program
  - **and here YOU are Brother!!!! :P**
- herding cats is much like, well, like herding cats.. 
- deciding the decider needs, well, needs the decider
- complaining = volunteering

Didn't see a doc link, but once I'm done being off-task, will build (regretably) on W10/VS2022 bare metal, then work my way backwards on VMs to VSPaleoEra10 .. haha... failing deciphering the obvious readme & debug logs, will nuisance up the board

THANK YOU FOR YOUR TIME TJ!
_~ Resident Village Idiot_

**Reply by Ted-juh** [reply, marked as answer]
While Ctrlr 5.3.201 is the "stable" version, I can only advice to aim at version 5.6.23. Vst2 is out of the window, Juce was updated to version 6 (prior version 5) and meant a huge overhaul for Ctrlr which Roman was, thankfully, providing us.

All changes made ( mouse events, vst3 name change support, and so on) will be made on the newer version anyway.

See Damien's fork CtlrX. Some excellent additions there.

**Reply by unityconstruct** [reply]
nice..! thx again for info.. I'd conduct my due diligence.. I have a working linux build, but VS makes for a smoother experience
I'm a tenacious hack, so what I lack in brains I normally resolve via attrition

**Reply by unityconstruct** [reply]
Well neither I nor any animals were injured in the process, so its a start! Thx for pointers @Ted-juh. 
Going to mark this answered now that I'm way off the original topic. THX AGAIN SIR!

- Guide made most the unobvious - obvious.
- I'm already on VS2022, so had to pull VS2019 Build Tools _(v142)_
- I initially tried CtrlrX, but was using JUCE7 & things puked horribly, so bailed on that & reread your tips & eventually found the mystical v5.6.0 on master... after checking commits to see the tag starting back at me.
- Process was smooth 2nd time around

## 2 notes-to-self: ##
- versions matter: clone & ensure on `master:5.6.0:3f9819e` commit
- JUCE 6.0.8 is needed (presumably) & Global dirs should point to WITHIN the ctrlr project

# More Notes to Self
###  WinGuide
https://godlike.com.au/fileadmin/godlike/techtools/ctrlr/guides/Compiling_Ctrlr_for_Windows_10_v2.1.pdf

### VS2022
- Universal Windows Platform development
- Desktop development with C++
- VS 2019 Build Tools (v142) [if using VS2022]

### Ctrlr Code
- on GitHub
  - https://github.com/damiensellier/CtrlrX/releases/tag/5.6.0
  - master: commit: 3f9819e
  - looks like this commit tagged 5.6.0
- Clone to `D:\source\repos\`
```bash
cd D:\source\repos\
git clone https://github.com/RomanKubiak/ctrlr
```
- Why not just create a new branch now? `master-init`
- unzip `D:\source\repos\ctrlr\Source\Misc\boost\boost.zip` -> `*\boost`

###  JUCE 6.0.8
- Ctrlr 5.6.0 appears to use JUCE 6.0.8
  - https://github.com/juce-framework/JUCE/releases/tag/6.0.8
  - so in order for JUCE to correctly parse things, the versions MUST MATCH
  - V7 induced syntax errors in the 1000s, with missing semi colons, etc
- I don't like installing ANYTHING, so pulled the ZIP instead of installer
  - Extracted to `D:\Program Files\JUCE608`
- Start with `D:\Program Files\JUCE608\Projucer.exe`
- Globals
  - Prog Path: `D:\source\repos\ctrlr\JUCE`
  - Modules Path: `D:\source\repos\ctrlr\JUCE\modules`
- File>Open: `D:\source\repos\ctrlrCtrlr.jucer`
  - If JUCE throws an `Invalid JUCE Path` message... likely you are not using correct JUCE version
    - double-check the Globals Path
    - check JUCE version is `6.0.8` (as of this note - based off ctrlr:master:5.6.0:3f9819e)
  - A drop down list showing `Visual Studio 2019` should be shown
    - There's no EXPORT button or anything helpful like that
    - so click the VS icon & pray VS2022(or 2019) starts loading

###  VS2022
- on first load, if VS version mismatch found, DO NOT update projects (requires Build Tools VS2019 (v142))
    - leave unchanged
- Init SLN Config
  - Build: Release | x64
- INIT PROJ CONFIG: `Ctrlr_SharedCode`
  - at this point, only need to config `Ctrlr_SharedCode` - so select it
  - `Project>Properties>Advanced>Preferred Build Tool Architecture`: 64-bit (x64)
  - `Project Properties>Configuration Properties>C/C++>OutputFiles`:
    - FROM: `$(IntDir)\`
    - TO: `<inherit from parent or project defaults>`
- `Ctrlr_SharedCode`>Build
```output
1>Done building project "Ctrlr_SharedCode.vcxproj".
========== Build: 1 succeeded, 0 failed, 0 up-to-date, 0 skipped ==========
========== Build completed at 3:11 PM and took 03:25.338 minutes ==========
```
VS2022 proj update msg
![cap 2024-0128-1400-32](https://github.com/RomanKubiak/ctrlr/assets/40268430/bd905bb0-c538-4c64-b609-8d9cc0953b26)
![cap 2024-0128-1400-41](https://github.com/RomanKubiak/ctrlr/assets/40268430/70280192-9214-4530-bab3-13401fb3d54e)
Pref Build Tool Arch(Note the project name at top)
![cap 2024-0128-1405-34](https://github.com/RomanKubiak/ctrlr/assets/40268430/09f1e4a7-f3ef-42c1-a2ef-33cec7eb38fb)
![cap 2024-0128-1405-42](https://github.com/RomanKubiak/ctrlr/assets/40268430/a1dc21d3-125f-4bb9-ac65-3530686ffd00)
Output files(Note the project name at top)
![cap 2024-0128-1406-15](https://github.com/RomanKubiak/ctrlr/assets/40268430/38ee85a2-ccbb-4ad6-8a53-9b727773920d)
![cap 2024-0128-1406-29](https://github.com/RomanKubiak/ctrlr/assets/40268430/a62a6b47-f007-43c0-8f50-fae5cb906c91)

**Reply by unityconstruct** [reply]
I love files... but not too much.. 260+ changes after build
.gitignore
```gitignore
**/VisualStudio2017/*
**/VisualStudio2019/*
```
now just 13

**Reply by unityconstruct** [reply]
for building `Ctrlr_SharedCode` so far
- JUCE open>export required only once
- assuming it just wires things up for VS the first time around.
- able to just start with VS2022 afterwards & building still happy

**Reply by unityconstruct** [reply]
- All 4 projects build success and Project>Properties config is the same for each
- Trapped observations
- - reminder: JUCE =>
  - `D:\source\repos\ctrlr\JUCE`
  - `D:\source\repos\ctrlr\JuceLibraryCode`
- reminder: lua => 
  - `D:\source\repos\ctrlr\Source\Misc\luabind\src`
  - `D:\source\repos\ctrlr\Source\Misc\lua\src`

<hr/>

# Visual Studio 2022
- on first load, if VS version mismatch found, DO NOT update projects (requires Build Tools VS2019 (v142))
    - leave unchanged
- Init SLN Config
  - Build: Release | x64

###  VS2022 PROJECT: Ctrlr_SharedCode
- INIT PROJ CONFIG: `Ctrlr_SharedCode`
  - at this point, only need to config `Ctrlr_SharedCode` - so select it
  - `Project>Properties>Advanced>Preferred Build Tool Architecture`: 64-bit (x64)
  - `Project Properties>Configuration Properties>C/C++>OutputFiles`:
    - FROM: `$(IntDir)\`
    - TO: `<inherit from parent or project defaults>`
- `Ctrlr_SharedCode`>Build

#### Build Notes
- generates: `D:\source\repos\ctrlr\Builds\VisualStudio2019\x64\Release\Shared Code\Ctrlr-x64.lib`
```output
1>Ctrlr_SharedCode.vcxproj -> D:\source\repos\ctrlr\Builds\VisualStudio2019\x64\Release\Shared Code\Ctrlr-x64.lib
1>Done building project "Ctrlr_SharedCode.vcxproj".
========== Build: 1 succeeded, 0 failed, 0 up-to-date, 0 skipped ==========
========== Build completed at 5:40 PM and took 03:32.808 minutes ==========
```

###  VS2022 PROJECT: Ctrlr_StandalonePlugin
- INIT PROJ CONFIG: `Ctrlr_StandalonePlugin`
  - at this point, only need to config `Ctrlr_StandalonePlugin` - so select it
  - `Project>Properties>Advanced>Preferred Build Tool Architecture`: 64-bit (x64)
  - `Project Properties>Configuration Properties>C/C++>OutputFiles`:
    - FROM: `$(IntDir)\`
    - TO: `<inherit from parent or project defaults>`
- `Ctrlr_StandalonePlugin`>Build

#### Build Notes
- JUCE lib: `include_juce_audio_plugin_client_Standalone.obj`
  - didn't see this in the build log, but didn't look hard either
- generates:  `D:\source\repos\ctrlr\Builds\VisualStudio2019\x64\Release\Standalone Plugin\Ctrlr-x64.exe`
- may or maynot have dependencies on `Ctrlr-x64.lib`, but based on size it seems the exe is truly `standalone`
- the build process may depend on the lib (rather than rebuild it blindly for each referencing project)
```output
1>Ctrlr_StandalonePlugin.vcxproj -> D:\source\repos\ctrlr\Builds\VisualStudio2019\x64\Release\Standalone Plugin\Ctrlr-x64.exe
1>Done building project "Ctrlr_StandalonePlugin.vcxproj".
========== Build: 1 succeeded, 0 failed, 1 up-to-date, 0 skipped ==========
========== Build completed at 5:51 PM and took 07:17.676 minutes ==========
```

###  VS2022 PROJECT:  Ctrlr_VST
- INIT PROJ CONFIG: `Ctrlr_VST`
  - at this point, only need to config `Ctrlr_VST` - so select it
  - `Project>Properties>Advanced>Preferred Build Tool Architecture`: 64-bit (x64)
  - `Project Properties>Configuration Properties>C/C++>OutputFiles`:
    - FROM: `$(IntDir)\`
    - TO: `<inherit from parent or project defaults>`
- `Ctrlr_VST`>Build

#### Build Notes
- JUCE lib: `include_juce_audio_plugin_client_VST2.cpp`
- generates: `D:\source\repos\ctrlr\Builds\VisualStudio2019\x64\Release\VST\Ctrlr-x64.dll`
```output
Started at 5:27 PM...
1>------ Build started: Project: Ctrlr_VST, Configuration: Release x64 ------
1>include_juce_audio_plugin_client_VST2.cpp
1>   Creating library D:\source\repos\ctrlr\Builds\VisualStudio2019\x64\Release\VST\Ctrlr-x64.lib and object D:\source\repos\ctrlr\Builds\VisualStudio2019\x64\Release\VST\Ctrlr-x64.exp
1>Generating code
1>D:\source\repos\ctrlr\Source\Core\CtrlrInlineUtilitiesGUI.h(87): warning C4715: 'gui::colourSchemeFromProperty': not all control paths return a value
1>D:\source\repos\ctrlr\Source\MIDI\CtrlrMidiInputComparatorSingle.cpp(239): warning C4702: unreachable code
1>Finished generating code
1>Ctrlr_VST.vcxproj -> D:\source\repos\ctrlr\Builds\VisualStudio2019\x64\Release\VST\Ctrlr-x64.dll
1>Done building project "Ctrlr_VST.vcxproj".
========== Build: 1 succeeded, 0 failed, 1 up-to-date, 0 skipped ==========
========== Build completed at 5:35 PM and took 07:40.480 minutes ==========
```

###  VS2022 PROJECT: Ctrlr_VST3
- INIT PROJ CONFIG: `Ctrlr_VST3`
  - at this point, only need to config `Ctrlr_VST3` - so select it
  - `Project>Properties>Advanced>Preferred Build Tool Architecture`: 64-bit (x64)
  - `Project Properties>Configuration Properties>C/C++>OutputFiles`:
    - FROM: `$(IntDir)\`
    - TO: `<inherit from parent or project defaults>`
- `Ctrlr_VST3`>Build
- 
#### Build Notes
- JUCE lib: `include_juce_audio_plugin_client_VST3.cpp`
- generates: `D:\source\repos\ctrlr\Builds\VisualStudio2019\x64\Release\VST3\Ctrlr-x64.vst3`

```output
Build started at 5:59 PM...
1>------ Build started: Project: Ctrlr_VST3, Configuration: Release x64 ------
1>include_juce_audio_plugin_client_VST3.cpp
1>   Creating library D:\source\repos\ctrlr\Builds\VisualStudio2019\x64\Release\VST3\Ctrlr-x64.lib and object D:\source\repos\ctrlr\Builds\VisualStudio2019\x64\Release\VST3\Ctrlr-x64.exp
1>Generating code
1>D:\source\repos\ctrlr\Source\Core\CtrlrInlineUtilitiesGUI.h(87): warning C4715: 'gui::colourSchemeFromProperty': not all control paths return a value
1>D:\source\repos\ctrlr\Source\MIDI\CtrlrMidiInputComparatorSingle.cpp(239): warning C4702: unreachable code
1>Finished generating code
1>Ctrlr_VST3.vcxproj -> D:\source\repos\ctrlr\Builds\VisualStudio2019\x64\Release\VST3\Ctrlr-x64.vst3
1>Done building project "Ctrlr_VST3.vcxproj".
========== Build: 1 succeeded, 0 failed, 1 up-to-date, 0 skipped ==========
========== Build completed at 6:07 PM and took 07:43.849 minutes ==========
```
